typing/typedecl_variance.ml

```diff
@@ -242,15 +212,9 @@ let compute_variance_type env ~check (required, loc) decl tyl =
         else (false, false) (* only check *)
       and i = concr  || i && tr = Private in
       let v = union v (make p n i) in
-      let v =
-        if not concr then v else
-        if mem Pos v && mem Neg v then full else
-        if Btype.is_Tvar ty then v else
-        union v
-          (if p then if n then full else covariant else conjugate covariant)
-      in
-      if decl.type_kind = Type_abstract && tr = Public then v else
-      set May_weak (mem May_neg v) v)
+      if not concr || Btype.is_Tvar ty then v else
+      union v
+        (if p then if n then full else covariant else conjugate covariant))
     params required
```

Comment:
`List.map2` returns the inferred variance of the parameters. For private types and constrained parameters, the variance is just the one given by the user, otherwise we use the inferred variance. The last line is for constrained type parameters in concrete datatypes. It is unchanged, but I'm wondering whether it is really needed, or correct. If a type is concrete, then it is injective, but should it really imply that there is really an occurrence?